pkg/apps/opssight/opssight.go

```diff
@@ -0,0 +1,256 @@
+}
+
+// NewOpsSight will return an OpsSight type
+func NewOpsSight(config *protoform.Config, kubeConfig *rest.Config) *OpsSight {
+	// Initialiase the clienset
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil
+	}
+	// Initialize the OpsSight client
+	opsSightClient, err := opssightclientset.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil
+	}
+	// Initialize Security Client
+	osSecurityClient, err := securityclient.NewForConfig(kubeConfig)
+	if err != nil {
+		osSecurityClient = nil
+	} else {
+		_, err := util.GetOpenShiftSecurityConstraint(osSecurityClient, "anyuid")
+		if err != nil {
+			osSecurityClient = nil
+		}
+	}
+	// Initialize the Route Client for Openshift routes
+	routeClient := util.GetRouteClient(kubeConfig)
+
+	// Initialize Black Duck Client
+	blackduckClient, err := blackduckclientset.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil
+	}
+	// Initialize creaters for different versions of OpsSight (each Creater can support differernt versions)
+	creaters := []Creater{
+		latestopssight.NewCreater(config, kubeConfig, kubeClient, opsSightClient, osSecurityClient, routeClient, blackduckClient),
+	}
+
+	return &OpsSight{
+		config:         config,
+		kubeConfig:     kubeConfig,
+		kubeClient:     kubeClient,
+		opsSightClient: opsSightClient,
+		routeClient:    routeClient,
+		creaters:       creaters,
+	}
+}
+
+// getCreater loops through each Creater and returns the one
+// that supports the specified version
+func (o OpsSight) getCreater(version string) (Creater, error) {
+	for _, c := range o.creaters {
+		for _, v := range c.Versions() {
+			if strings.Compare(v, version) == 0 {
+				return c, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("version %s is not supported", version)
+}
+
+// Versions returns the versions that the operator supports for OpsSight
+func (o OpsSight) Versions() []string {
+	var versions []string
+	// Get versions that each Creater supports
+	for _, c := range o.creaters {
+		for _, v := range c.Versions() {
+			versions = append(versions, v)
+		}
+	}
+	return versions
+}
+
+// Ensure will get the necessary Creater and make sure the instance
+// is correctly deployed or deploy it if needed
+func (o OpsSight) Ensure(ops *opssightapi.OpsSight) error {
+	// If the version is not specified then we set it to be the latest.
+	log.Debugf("ensure OpsSight %s", ops.Spec.Namespace)
+	if len(ops.Spec.Version) == 0 {
+		versions := o.Versions()
+		sort.Sort(sort.Reverse(sort.StringSlice(versions)))
+		ops.Spec.Version = versions[0]
+		log.Debugf("setting OpsSight version to %s", ops.Spec.Version)
+	}
+
+	creater, err := o.getCreater(ops.Spec.Version) // get Creater for the OpsSight Version
+	if err != nil {
+		return fmt.Errorf("failed to get OpsSight creater for version %s: %s", ops.Spec.Version, err)
+	}
+	return creater.Ensure(ops) // Ensure the OpsSight
+}
+
+// Delete will delete the Black Duck OpsSight
+func (o OpsSight) Delete(namespace string) error {
```

Comment:
@msenmurugan @mphammer  Shouldn't this function delete Prometheus too?

Reply:
Yup. It needs too.. It is already deleting... All the objects are under app=opssight @jdartigalongue

Reply:
@msenmurugan It currently deletes all the replication controllers with label 'app=opssight'  but prometheus is a deployment and not a RC.

Reply:
@msenmurugan @jdartigalongue I made it so if there are Replication Controllers that do not belong to OpsSight in the namespace, then Delete() will delete all Replication Controllers **and** Deployments that have the label "app=opssight" from the namespace.